File: typescript/src/exceptions.ts
/** Base exception for all Vault SDK errors. */
export class VaultError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "VaultError";
  }
}

/** Sentinel is unreachable or the request timed out. */
export class VaultConnectionError extends VaultError {
  constructor(message: string) {
    super(message);
    this.name = "VaultConnectionError";
  }
}

/** Sentinel returned a non-2xx response. */
export class VaultHttpError extends VaultError {
  /** HTTP status code from Sentinel. */
  public readonly statusCode: number;
  /** Error detail message from Sentinel's response body. */
  public readonly detail: string;

  constructor(statusCode: number, detail: string) {
    super(`HTTP ${statusCode}: ${detail}`);
    this.name = "VaultHttpError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
}

/** 404 — requested resource does not exist. */
export class VaultNotFoundError extends VaultHttpError {
  constructor(detail: string = "Not found") {
    super(404, detail);
    this.name = "VaultNotFoundError";
  }
}

/** 401/403 — invalid or missing internal key. */
export class VaultAuthError extends VaultHttpError {
  constructor(statusCode: number, detail: string = "Unauthorized") {
    super(statusCode, detail);
    this.name = "VaultAuthError";
  }
}

File: python/hermes_vault/exceptions.py
class VaultError(Exception):
    """Base exception for all Vault SDK errors."""


class VaultConnectionError(VaultError):
    """Sentinel is unreachable or request timed out."""


class VaultHTTPError(VaultError):
    """Sentinel returned a non-2xx response."""

    def __init__(self, status_code: int, detail: str):
        self.status_code = status_code
        self.detail = detail
        super().__init__(f"HTTP {status_code}: {detail}")


class VaultNotFoundError(VaultHTTPError):
    """404 — requested resource does not exist."""

    def __init__(self, detail: str = "Not found"):
        super().__init__(404, detail)


class VaultAuthError(VaultHTTPError):
    """401/403 — invalid or missing internal key."""

    def __init__(self, status_code: int, detail: str = "Unauthorized"):
        super().__init__(status_code, detail)